File: app/src/test/java/ch/threema/client/NonceFactoryTest.java
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema Java Client
 * Copyright (c) 2017-2021 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.client;
import static org.mockito.Mockito.*;

import ch.threema.base.ThreemaException;
import org.junit.Assert;
import org.junit.Test;

import java.security.SecureRandom;

public class NonceFactoryTest {

	// Hack to "Mock" the Secure Random
	private class SecureRandomMocken extends SecureRandom {
		private byte[][] next;
		private int nextPos = 0;

		public void nextNonces(byte[]... next) {
			this.next = next;
		}

		@Override
		public void nextBytes(byte[] bytes) {
			// Check length
			Assert.assertEquals(bytes.length, 24);

			if (this.next != null && this.nextPos < this.next.length) {
				for(int n = 0; n < bytes.length; n++) {
					bytes[n] = this.next[this.nextPos][n];
				}
				this.nextPos++;
				return;
			}

			super.nextBytes(bytes);
		}
	}

	@Test
	public void testNext() throws Exception {
		NonceStoreInterface nonceStoreMock = mock(NonceStoreInterface.class);
		SecureRandomMocken secureRandomMock = new SecureRandomMocken();

		// Store always return true
		when(nonceStoreMock.store(any())).thenReturn(true);

		NonceFactory factory = new NonceFactory(secureRandomMock, nonceStoreMock);
		byte[] result = factory.next();

		// Check if store is called
		verify(nonceStoreMock, times(1)).store(any());

		// Verify the result
		Assert.assertEquals(24, result.length);
	}

	@Test
	public void testNext2Times() throws Exception {
		NonceStoreInterface nonceStoreMock = mock(NonceStoreInterface.class);
		SecureRandomMocken secureRandomMock = new SecureRandomMocken();

		byte[] existingNonce = new byte[]{0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,
				0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,
				0x01,0x01,0x01,0x01};

		byte[] newNonce = new byte[]{0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,
				0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,
				0x01,0x01,0x01,0x02};

		// Store always return true
		when(nonceStoreMock.store(eq(existingNonce))).thenReturn(false);
		when(nonceStoreMock.store(eq(newNonce))).thenReturn(true);

		NonceFactory factory = new NonceFactory(secureRandomMock, nonceStoreMock);
		secureRandomMock.nextNonces(existingNonce, newNonce);
		byte[] result = factory.next();

		// Check if store is called twice
		verify(nonceStoreMock, times(2)).store(any());

		// Verify the result
		Assert.assertEquals(24, result.length);
	}


	@Test(expected=ThreemaException.class)
	public void testNextXTimes() throws Exception {
		NonceStoreInterface nonceStoreMock = mock(NonceStoreInterface.class);

		// Store always return true
		when(nonceStoreMock.store(any())).thenReturn(false);

		NonceFactory factory = new NonceFactory(new SecureRandom(), nonceStoreMock);
		factory.next();
	}

	@Test
	public void testNextWithoutStore() throws Exception {
		NonceStoreInterface nonceStoreMock = mock(NonceStoreInterface.class);

		NonceFactory factory = new NonceFactory(new SecureRandom(), nonceStoreMock);
		factory.next(false);

		verify(nonceStoreMock, never()).store(any());
	}

	@Test
	public void textNextWithoutStore() throws Exception {
		byte[] existingNonce = new byte[]{0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,
				0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,0x01,
				0x01,0x01,0x01,0x01};

		NonceStoreInterface nonceStoreMock = mock(NonceStoreInterface.class);
		when(nonceStoreMock.exists(eq(existingNonce))).thenReturn(true);
		NonceFactory factory = new NonceFactory(new SecureRandom(), nonceStoreMock);
		Assert.assertTrue(factory.exists(existingNonce));


		when(nonceStoreMock.exists(eq(existingNonce))).thenReturn(false);
		Assert.assertFalse(factory.exists(existingNonce));
	}

}

File: domain/src/test/java/ch/threema/base/crypto/NonceFactoryTest.kt
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2017-2025 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.base.crypto

import ch.threema.base.utils.SecureRandomUtil.generateRandomBytes
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import java.util.LinkedList
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class NonceFactoryTest {
    @Test
    fun testNext() {
        val nonceStoreMock = mockk<NonceStore>(relaxed = true)

        val nonce = Nonce(
            byteArrayOf(
                0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
                0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
                0x01, 0x01, 0x01, 0x01,
            ),
        )

        val nonceProvider = TestNonceProvider()
        nonceProvider.addNextNonces(listOf(nonce.bytes))

        // Store always return true
        every { nonceStoreMock.store(any<NonceScope>(), nonce) } returns true
        val factory = NonceFactory(nonceStoreMock, nonceProvider)
        val result: Nonce = factory.next(NonceScope.CSP)

        // Check if uniqueness is verified
        verify(exactly = 1) { nonceStoreMock.exists(any<NonceScope>(), nonce) }

        // Verify the result
        assertEquals(24, result.bytes.size)
    }

    @Test
    fun testNext2Times() {
        val nonceProvider = TestNonceProvider()
        val existingNonce = byteArrayOf(
            0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
            0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
            0x01, 0x01, 0x01, 0x01,
        )
        val newNonce = byteArrayOf(
            0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
            0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
            0x01, 0x01, 0x01, 0x02,
        )

        nonceProvider.addNextNonces(listOf(existingNonce, newNonce))

        val nonceStoreMock = mockk<NonceStore>()
        every { nonceStoreMock.exists(any<NonceScope>(), Nonce(existingNonce)) } returns true
        every { nonceStoreMock.exists(any<NonceScope>(), Nonce(newNonce)) } returns false

        val factory = NonceFactory(nonceStoreMock, nonceProvider)
        val result: Nonce = factory.next(NonceScope.CSP)

        // Check if uniqueness is verified twice (because the first nonce already existed)
        verify(exactly = 1) { nonceStoreMock.exists(any<NonceScope>(), Nonce(existingNonce)) }
        verify(exactly = 1) { nonceStoreMock.exists(any<NonceScope>(), Nonce(newNonce)) }

        // Verify the result
        assertEquals(24, result.bytes.size)
    }

    @Test
    fun testNextDoesNotStoreNonce() {
        val nonceStoreMock = mockk<NonceStore>(relaxed = true)

        val factory = NonceFactory(nonceStoreMock)
        val createdNonce = factory.next(NonceScope.CSP)

        verify(exactly = 0) { nonceStoreMock.store(any<NonceScope>(), createdNonce) }
    }

    @Test
    fun testExistsReturnsValueReturnedByNonceStore() {
        val existingNonce = Nonce(
            byteArrayOf(
                0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
                0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
                0x01, 0x01, 0x01, 0x01,
            ),
        )

        val nonceStoreMock = mockk<NonceStore>()

        every { nonceStoreMock.exists(any<NonceScope>(), existingNonce) } answers {
            val nonce = secondArg<ByteArray>()
            nonce.contentEquals(existingNonce.bytes)
        }

        val factory = NonceFactory(nonceStoreMock)
        assertTrue(factory.exists(NonceScope.CSP, existingNonce))

        every { nonceStoreMock.exists(any<NonceScope>(), existingNonce) } answers {
            val nonce = secondArg<ByteArray>()
            !nonce.contentEquals(existingNonce.bytes)
        }

        assertFalse(factory.exists(NonceScope.CSP, existingNonce))
    }
}

private class TestNonceProvider : NonceFactoryNonceBytesProvider {
    private val nextNonces = LinkedList<ByteArray>()
    fun addNextNonces(nonces: List<ByteArray>) {
        nextNonces.addAll(nonces)
    }

    override fun next(length: Int): ByteArray {
        // Check length
        assertEquals(24, length.toLong())
        val nonce: ByteArray? = nextNonces.pollFirst()
        if (nonce != null) {
            assertEquals(24, nonce.size.toLong())
        }
        return nonce ?: generateRandomBytes(length)
    }
}
